miio/push_server/serverprotocol.py

```diff
@@ -52,57 +52,101 @@ def send_ping_ACK(self, host, port):
         self.transport.sendto(m, (host, port))
         _LOGGER.debug("%s:%s<=ACK(server_id=%s)", host, port, self.server.server_id)
 
-    def send_msg_OK(self, host, port, msg_id, token):
-        # This result means OK, but some methods return ['ok'] instead of 0
-        # might be necessary to use different results for different methods
-        result = {"result": 0, "id": msg_id}
+    def _create_message(self, data, token, device_id):
+        """Create a message to be sent to the client."""
         header = {
             "length": 0,
             "unknown": 0,
-            "device_id": self.server.server_id,
+            "device_id": device_id,
             "ts": datetime.datetime.now(),
         }
         msg = {
-            "data": {"value": result},
+            "data": {"value": data},
             "header": {"value": header},
             "checksum": 0,
         }
         response = Message.build(msg, token=token)
-        self.transport.sendto(response, (host, port))
+
+        return response
+
+    def send_response(self, host, port, msg_id, token, payload=None):
+        if payload is None:
+            payload = {}
+
+        result = {**payload, "id": msg_id}
+        msg = self._create_message(result, token, device_id=self.server.server_id)
+
+        self.transport.sendto(msg, (host, port))
         _LOGGER.debug(">> %s:%s: %s", host, port, result)
 
-    def datagram_received(self, data, addr):
-        """Handle received messages."""
-        try:
-            (host, port) = addr
-            if data == HELO_BYTES:
-                self.send_ping_ACK(host, port)
-                return
+    def send_error(self, host, port, msg_id, token, code, message):
+        """Send error message with given code and message to the client."""
+        return self.send_response(
+            host, port, msg_id, token, {"error": {"code": code, "error": message}}
+        )
+
+    def _handle_datagram_from_registered_device(self, host, port, data):
+        """Handle requests from registered eventing devices."""
+        token = self.server._registered_devices[host]["token"]
+        callback = self.server._registered_devices[host]["callback"]
+
+        msg = Message.parse(data, token=token)
+        msg_value = msg.data.value
+        msg_id = msg_value["id"]
+        _LOGGER.debug("<< %s:%s: %s", host, port, msg_value)
 
-            if host not in self.server._registered_devices:
```

Comment:
I would like to have this error back when the host is not registered as a miio_device or a client.
See comment below.

Reply:
This is now checked inside `datagram_received`, for registered devices. But you are right, we are responding to all non-registered devices as if they were real clients..